shared/fs/row/rows-container.js

```diff
@@ -5,163 +5,164 @@ import * as Types from '../../constants/types/fs'
+)
+
+const _getPlaceholderRows = (type): I.List<RowTypes.PlaceholderRowItem> =>
+  I.List([
+    {key: 'placeholder:1', name: '1', rowType: 'placeholder', type},
+    {key: 'placeholder:2', name: '2', rowType: 'placeholder', type},
+    {key: 'placeholder:3', name: '3', rowType: 'placeholder', type},
+  ])
+const filePlaceholderRows = _getPlaceholderRows('file')
+const folderPlaceholderRows = _getPlaceholderRows('folder')
+
+const _makeInTlfRows = memoize((editingRows, amendedStillRows) => editingRows.concat(amendedStillRows))
+
+const getInTlfItemsFromStateProps = (stateProps, path: Types.Path): I.List<RowTypes.NamedRowItem> => {
   const _pathItem = stateProps._pathItems.get(path, Constants.unknownPathItem)
   if (_pathItem.type !== 'folder') {
-    return getPlaceholderRows('file')
+    return filePlaceholderRows
   }
 
   if (_pathItem.progress === 'pending') {
-    return getPlaceholderRows('file')
+    return filePlaceholderRows
   }
 
   const editingRows = getEditingRows(stateProps._edits, path)
-  const stillRows = getStillRows(stateProps._pathItems, path, _pathItem.children.toArray())
+  const stillRows = getStillRows(stateProps._pathItems, path, _pathItem.children)
 
   return sortRowItems(
-    editingRows.concat(amendStillRows(stillRows, stateProps._uploads)),
+    _makeInTlfRows(editingRows, amendStillRowsWithUploads(stillRows, stateProps._uploads)),
     stateProps._sortSetting,
-    undefined
+    ''
   )
 }
 
-const getRootRows = stateProps =>
-  sortRowItems(
-    [
-      {key: 'tlfType:private', name: 'private', rowType: 'tlf-type', type: 'folder'},
-      {key: 'tlfType:public', name: 'public', rowType: 'tlf-type', type: 'folder'},
-      {key: 'tlfType:team', name: 'team', rowType: 'tlf-type', type: 'folder'},
-    ],
-    stateProps._sortSetting,
-    undefined
-  )
-
-const getTlfRowsFromTlfs = (tlfs: I.Map<string, Types.Tlf>, tlfType: Types.TlfType): Array<SortableRowItem> =>
-  tlfs.reduce(
-    (rows, {isIgnored, isNew}, name) =>
-      isIgnored
-        ? rows
-        : [
-            ...rows,
-            {
-              isNew,
-              key: `tlf:${name}`,
-              name,
-              rowType: 'tlf',
-              tlfType,
-              type: 'folder',
-            },
-          ],
-    ([]: Array<SortableRowItem>)
-  )
-
-const getTlfItemsFromStateProps = (stateProps, path) => {
+const getRootRows = (stateProps): I.List<RowTypes.TlfTypeRowItem> =>
+  I.List([
+    {key: 'tlfType:private', name: 'private', rowType: 'tlf-type', type: 'folder'},
+    {key: 'tlfType:public', name: 'public', rowType: 'tlf-type', type: 'folder'},
+    {key: 'tlfType:team', name: 'team', rowType: 'tlf-type', type: 'folder'},
+  ])
+
+const getTlfRowsFromTlfs = memoize(
+  (tlfs: I.Map<string, Types.Tlf>, tlfType: Types.TlfType): I.List<SortableRowItem> =>
+    I.List().withMutations(list =>
+      tlfs.reduce(
+        (rows, {isIgnored, isNew}, name) =>
+          isIgnored
+            ? rows
+            : rows.push({
```

Comment:
Doesn't this return the length? I think you want `rows.concat`, or the spread operator like it was before.

Reply:
Ha; took me a while to understand how this code works ... this is `I.List` now, not and `Array`. So `.push` actually returns a new list: https://immutable-js.github.io/immutable-js/docs/#/List/push